tensorboard/data/server/run.rs

```diff
@@ -0,0 +1,363 @@
+    /// with actual start time. See [`EventFile::Dead`] for conditions under which an event file
+    /// may be dead. Once an event file is added to this map, it may become dead, but it will not
+    /// be removed entirely. This way, we know not to just re-open it again at the next load cycle.
+    files: BTreeMap<PathBuf, EventFile<BufReader<File>>>,
+
+    /// Reservoir-sampled data and metadata for each time series.
+    time_series: HashMap<Tag, TimeSeries>,
+}
+
+#[derive(Debug)]
+enum EventFile<R> {
+    /// An event file that may still have more valid data.
+    Active(EventFileReader<R>),
+    /// An event file that can no longer be read.
+    ///
+    /// This can be because of a non-recoverable read error (e.g., a bad length checksum), due to
+    /// the last-read record being very old (note: not yet implemented), or due to the file being
+    /// deleted.
+    Dead,
+}
+
+#[derive(Debug)]
+struct TimeSeries {
+    data_class: pb::DataClass,
+    metadata: Box<pb::SummaryMetadata>,
+    rsv: StageReservoir<StageValue>,
+}
+
+impl TimeSeries {
+    fn new(metadata: Box<pb::SummaryMetadata>) -> Self {
+        let data_class =
+            pb::DataClass::from_i32(metadata.data_class).unwrap_or(pb::DataClass::Unknown);
+        let capacity = match data_class {
+            pb::DataClass::Scalar => 1000,
+            pb::DataClass::Tensor => 100,
+            pb::DataClass::BlobSequence => 10,
+            _ => 0,
+        };
+        Self {
+            data_class,
+            metadata,
+            rsv: StageReservoir::new(capacity),
+        }
+    }
+}
+
+/// A value staged in the reservoir.
+///
+/// This is kept as close as possible to the on-disk event representation, since every record in
+/// the stream is converted into this format.
+#[derive(Debug)]
+struct StageValue {
+    wall_time: WallTime,
+    payload: StagePayload,
+}
+
+#[derive(Debug)]
+enum StagePayload {
+    #[allow(dead_code)]
+    GraphDef(Vec<u8>),
+    Summary(Box<pb::summary::value::Value>),
+}
+
+impl RunLoader {
+    pub fn new() -> Self {
+        Self {
+            start_time: None,
+            files: BTreeMap::new(),
+            time_series: HashMap::new(),
+        }
+    }
+
+    /// Loads new data given the current set of event files.
+    ///
+    /// The provided filenames should correspond to the entire set of event files currently part of
+    /// this run.
+    pub fn reload(&mut self, filenames: Vec<PathBuf>) {
+        self.update_file_set(filenames);
+        self.reload_files();
+    }
+
+    /// Updates the active key set of `self.files` to match the given filenames.
+    ///
+    /// After this function returns, `self.files` may still have keys not in `filenames`, but they
+    /// will all map to [`EventFile::Dead`].
+    fn update_file_set(&mut self, filenames: Vec<PathBuf>) {
+        // Remove any discarded files.
+        let new_file_set: HashSet<&PathBuf> = filenames.iter().collect();
```

Comment:
Of some interest: you have probably seen us write `.collect()` before to
turn an iterator into a vector. But here we’re turning one into a hash
set! In fact, [`collect`] can generate any collection type, as long as
it implements [`FromIterator`]. The specific type in any instance is
determined by context, and the compiler will complain if there’s not
enough information to uniquely identify a type. This is sometimes called
[target typing].

[`collect`]: https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.collect
[`FromIterator`]: https://doc.rust-lang.org/std/iter/trait.FromIterator.html
[target typing]: https://docs.oracle.com/javase/tutorial/java/javaOO/lambdaexpressions.html#target-typing